rust/kernel/sync/condvar.rs

```diff
@@ -0,0 +1,251 @@
+///     value_changed: CondVar,
+/// }
+///
+/// /// Waits for `e.value` to become `v`.
+/// fn wait_for_vaue(e: &Example, v: u32) {
+///     while e.value.load(Ordering::Acquire) != v {
+///         e.value_changed.wait_nolock_uninterruptible();
+///     }
+/// }
+///
+/// /// Increments `e.value` and notifies all potential waiters.
+/// fn increment(e: &Example) {
+///     e.value.fetch_add(1, Ordering::Release);
+///     e.value_changed.notify_all();
+/// }
+///
+/// /// Allocates a new boxed `Example`.
+/// fn new_example() -> Result<Pin<Box<Example>>> {
+///     Box::pin_init(pin_init!(Example {
+///         value: AtomicU32::new(0),
+///         value_changed <- new_condvar!("Example::value_changed"),
+///     }))
+/// }
+/// ```
+///
+/// [`struct wait_queue_head`]: ../../../include/linux/wait.h
+#[pin_data]
+pub struct CondVar {
+    #[pin]
+    pub(crate) wait_list: Opaque<bindings::wait_queue_head>,
+
+    /// A condvar needs to be pinned because it contains a [`struct list_head`] that is
+    /// self-referential, so it cannot be safely moved once it is initialised.
+    #[pin]
+    _pin: PhantomPinned,
+}
+
+// SAFETY: `CondVar` only uses a `struct wait_queue_head`, which is safe to use on any thread.
+#[allow(clippy::non_send_fields_in_send_ty)]
+unsafe impl Send for CondVar {}
+
+// SAFETY: `CondVar` only uses a `struct wait_queue_head`, which is safe to use on multiple threads
+// concurrently.
+unsafe impl Sync for CondVar {}
+
+impl CondVar {
+    /// Constructs a new condvar initialiser.
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(name: &'static CStr, key: &'static LockClassKey) -> impl PinInit<Self> {
+        pin_init!(Self {
+            _pin: PhantomPinned,
+            // SAFETY: `__init_waitqueue_head` initialises the waitqueue head, and both `name` and
+            // `key` have static lifetimes so they live indefinitely.
+            wait_list <- unsafe {
+                Opaque::ffi_init2(
+                    bindings::__init_waitqueue_head,
+                    name.as_char_ptr(),
+                    key.as_ptr(),
+                )
+            },
+        })
+    }
+
+    /// Atomically releases the given lock (whose ownership is proven by the guard) and puts the
+    /// thread to sleep. It wakes up when notified by [`CondVar::notify_one`] or
+    /// [`CondVar::notify_all`], or when the thread receives a signal.
+    ///
+    /// Returns whether there is a signal pending.
+    pub fn wait_internal<T, B>(&self, wait_state: u32, guard: Option<&mut Guard<'_, T, B>>)
+    where
+        T: ?Sized,
+        B: Backend,
+    {
+        let wait = Opaque::<bindings::wait_queue_entry>::uninit();
+
+        // SAFETY: `wait` points to valid memory.
+        unsafe { bindings::init_wait(wait.get()) };
+
+        // SAFETY: Both `wait` and `wait_list` point to valid memory.
+        unsafe {
+            bindings::prepare_to_wait_exclusive(self.wait_list.get(), wait.get(), wait_state as _)
+        };
```

Comment:
When `guard` is `None`, need to recheck the wait condition here, otherwise, a wakeup/notify can happen before the `prepare_to_wait_exclusive` above, and we still go to sleep due to not checking the wait condition. IOW, it's a missed wakeup. This is also why `____wait_event` checks the condition.

Maybe
```rust
pub fn wait_internal<T, B>(
  &self,
  wait_state: u32,
  guard: Option<&mut Guard<'_, T, B>,
  check: impl FnOnce(&Option<&mut Guad<'_,T, B>) -> bool)
where
        T: ?Sized,
        B: Backend
{
// ...
        unsafe {
            bindings::prepare_to_wait_exclusive(self.wait_list.get(), wait.get(), wait_state as _)
        };

        if (check(&guard)) {
            unsafe { bindings::finish_wait(self.wait_list.get(), wait.get()) };
            return;
        }
// ...
}
```

However, I think a better approach would be wrapping `wait_event` first.